### PR TITLE
add TFF post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+    round:  14   accuracy:  0.7788 
+    round:  15   accuracy:  0.7814 
+    round:  16   accuracy:  0.7836 
+    round:  17   accuracy:  0.7855 
+    round:  18   accuracy:  0.7872 
+    round:  19   accuracy:  0.7885 
+    round:  20   accuracy:  0.7902 
+
+Training accuracy is increasing continuously. These values represent averages of
+*local* accuracy measurements, so in the real world, might well be overly
```
Comment (batpigandme) [inline _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:447]:
```suggestion
*local* accuracy measurements, so in the real world, they might well be overly
```

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+optimistic (with each client overfitting on their respective data). So
+supplementing federated training, a federated evaluation process would need to
+be built in order to get a realistic view on performance. This is a topic to
+come back to when more related TFF documentation is available.
+
+### Conclusion
+
+We hope you've enjoyed this first introduction to TFF using R. Certainly at this
+time, it is too early for productive use; and for application in research (e.g., adversarial attacks on federated learning)
```
Comment (batpigandme) [inline _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:456]: By productive use, do you mean use in production? If yes, then I'd say use in production, but, if no, you're good. 🙂

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd [hunk 3]
```diff
@@ -0,0 +1,464 @@
+---
+title: "A first look at federated learning with TensorFlow"
+description: > 
+  The term "federated learning" was coined to describe a form of distributed model training where the data remains on the clients, i.e., is never shipped to the coordinating server. In this post, we introduce central concepts and run first experiments with TensorFlow Federated, using R.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:4]: not sure I get it completely- it's clients (several) vs. server (one) ... "clients" is supposed to be a standalone noun...
Now that I think about it: yeah, right, I should say "client devices"! thanks!

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+```
+
+Here, stereotypically, is the process of applied deep learning: Gather/get data;
+iteratively train and evaluate; deploy. Repeat (or have it all automated as a
+continuous workflow anyway). We often discuss training and evaluation;
+deployment matters to various degrees, depending on the circumstances. But the
+data often is just assumed to be there: All together, in one place. (On your
+laptop; on a central server; in some cluster in the cloud.) In real life though,
+data could be all over the world: on smartphones for example, or on IoT devices.
+There's a lot of reasons why we don't want to ship all that data to some central
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:32]: Interesting! I think I may be doing the "hypercorrect" thing sometimes, from a non-native speaker's point of view ... (like, I normally use "data" as singular, probably because at some point in the past I must have found that very strange)

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+server, and this is where privacy guarantees may be violated. In some cases, it
+may be easy to reconstruct the actual data from the gradients -- in an NLP task,
+for example, when the vocabulary is known on the server, and gradient updates
+are sent for small pieces of text.
+
+This may sound like a special case, but general methods have been demonstrated
+that work regardless of circumstances. For example, Zhu et
+al. [@abs-1906-08935] use a "generative" approach, with the server starting
+from randomly generated fake data (resulting in fake gradients) and then,
+iteratively updating that data to obtain gradients more and more like the real
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:109]: that should really be "updating" (taking up the previous "starting")

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+often achieved through [secret
+sharing](https://en.wikipedia.org/wiki/Secret_sharing), where individual pieces
+of data (e.g.: individual salaries) are split up into "shares", exchanged and
+combined with random data in various ways, until finally the desired global
+result (e.g.: mean salary) is computed. (These are extremely fascinating topics
+that unfortunately, by far surpass the scope of this post.)
+
+Now, with the server prevented from actually "seeing" the gradients, a problem
+still remains. The model -- especially a high-capacity one, with many parameters
+-- could still memorize individual training data. Here is where *differential
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:128]: yeah, it totally is, but thanks anyway!

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+}
+```
+
+Next is the call to `tff$tf_computation` we already alluded to, wrapping
+`get_local_temperature_average`. We also need to indicate the
+argument's TFF-level type. 
+(In the context of this post, TFF datatypes are
+definitely out-of-scope, but the TFF documentation has lots of detailed
+information on them (and not just on them, actually, but the design of TFF in
+general). All we need to know right now is that we will be able to pass the data
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:201]: yeah, right, I'll think of something...

File: _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd:456]: yeah, I do... wow, here I really would have thought that works, but doesn't seem like...

**Comment by skeydan** [comment]
Thanks a 1000 times Mara!!
